branches/doc_maggen/man/man3/utilities.3
.TH "utilities" 3 "4 Sep 2010" "Version 1.0" "maggen" \" -*- nroff -*-
.ad l
.nh
.SH NAME
utilities \- 
.SH SYNOPSIS
.br
.PP
.SS "Functions"

.in +1c
.ti -1c
.RI "bool \fBcreate_folder\fP (const string path)"
.br
.ti -1c
.RI "bool \fBclean_output_folder\fP (const string path)"
.br
.ti -1c
.RI "bool \fBcopy_static_code\fP (const string path_d, const string path_s)"
.br
.ti -1c
.RI "void \fBgenerate_names_instance\fP (const Graph &graph, string datas[], size_t size_d)"
.br
.ti -1c
.RI "void \fBgenerate_names_attr\fP (const Graph &graph, string datas[], size_t size_d)"
.br
.ti -1c
.RI "void \fBprint_graph\fP (const Graph &graph, const string path, const string name_file, const string name_graph, const string names[], string shape_vertex)"
.br
.ti -1c
.RI "Vertex \fBreturn_vertex\fP (const Graph &graph, const \fBExpr_leaf\fP *node)"
.br
.ti -1c
.RI "void \fBmerge_graph\fP (const Graph &graph1, const Graph &graph2, Graph &graph_merged)"
.br
.ti -1c
.RI "void \fBproject_graph\fP (const \fBSymbol\fP *symb, Graph &graph)"
.br
.ti -1c
.RI "string \fBcleaning_tabs\fP (const string str)"
.br
.ti -1c
.RI "string \fBwrite_inf_context\fP (const vector< unsigned short > &context_vec)"
.br
.ti -1c
.RI "bool \fBbelong_index\fP (const unsigned short &index, const vector< unsigned short > &vec)"
.br
.in -1c
.SH "Function Documentation"
.PP 
.SS "bool utilities::belong_index (const unsigned short & index, const vector< unsigned short > & vec)"Seachs a index in a vector. 
.PP
\fBParameters:\fP
.RS 4
\fIindex\fP 
.br
\fIvec\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP

.PP
Definition at line 211 of file Utilities.cpp.
.PP
Referenced by genevalmag::Builder_visit_sequences::gen_visit_seq(), genevalmag::Builder_visit_sequences::generate_visit_sequences(), genevalmag::plan_family_computed(), and genevalmag::purge_plan_with().
.SS "bool utilities::clean_output_folder (const string path)"Remove and create the output folder of files .dot and .png. 
.PP
\fBParameters:\fP
.RS 4
\fIpath\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP
Remove and create the output folder of files .dot and .png. 
.PP
Definition at line 41 of file Utilities.cpp.
.PP
References create_folder().
.PP
Referenced by genevalmag::Maglib::gen_evaluator(), genevalmag::Builder_graphs::save_adp_graphs(), genevalmag::Builder_graphs::save_all_graphs(), genevalmag::Builder_plans::save_all_plans(), genevalmag::Builder_plans::save_all_plans_project(), genevalmag::Builder_graphs::save_cyclic_graphs(), genevalmag::Builder_graphs::save_dcg_graphs(), genevalmag::Builder_graphs::save_down_graphs(), and genevalmag::Builder_graphs::save_dp_graphs().
.SS "string utilities::cleaning_tabs (const string str)"Remove tabs and replace for spaces. 
.PP
\fBParameters:\fP
.RS 4
\fIstr\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP
Remove tabs and replace for spaces. 
.PP
Definition at line 226 of file Utilities.cpp.
.PP
Referenced by genevalmag::create_instance(), genevalmag::Builder_code::generate_all_methods_eqs(), genevalmag::Builder_graphs::save_adp_graphs(), genevalmag::Builder_plans::save_all_plans(), genevalmag::Builder_plans::save_all_plans_project(), genevalmag::Builder_graphs::save_cyclic_graphs(), genevalmag::Builder_graphs::save_dcg_graphs(), genevalmag::Builder_graphs::save_dp_graphs(), and genevalmag::save_rvalue().
.SS "bool utilities::copy_static_code (const string path_d, const string path_s)"Copies the static file to the generation path. 
.PP
\fBParameters:\fP
.RS 4
\fIpath_d\fP 
.br
\fIpath_s\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP
Copies the static file to the generation path. 
.PP
Definition at line 57 of file Utilities.cpp.
.PP
Referenced by genevalmag::Builder_code::generate_code().
.SS "bool utilities::create_folder (const string path)"Create the folder passed as parameter. 
.PP
\fBParameters:\fP
.RS 4
\fIpath\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP
Create the folder passed as parameter. 
.PP
Definition at line 26 of file Utilities.cpp.
.PP
Referenced by clean_output_folder().
.SS "void utilities::generate_names_attr (const Graph & graph, string datas[], size_t size_d)"Generates the names of vertex for down graph. (The vertex's name is an attribute). 
.PP
\fBParameters:\fP
.RS 4
\fIgraph\fP 
.br
\fIdatas\fP 
.br
\fIsize_d\fP Generates the names of vertex for down graph. (The vertex's name is an attribute). 
.RE
.PP

.PP
Definition at line 88 of file Utilities.cpp.
.PP
Referenced by genevalmag::Builder_graphs::save_down_graphs().
.SS "void utilities::generate_names_instance (const Graph & graph, string datas[], size_t size_d)"Generates the names of vertex. (The vertex's name is an instance). 
.PP
\fBParameters:\fP
.RS 4
\fIgraph\fP 
.br
\fIdatas\fP 
.br
\fIsize_d\fP Generates the names of vertex. (The vertex's name is an instance). 
.RE
.PP

.PP
Definition at line 75 of file Utilities.cpp.
.PP
Referenced by genevalmag::Builder_graphs::save_adp_graphs(), genevalmag::Builder_graphs::save_cyclic_graphs(), genevalmag::Builder_graphs::save_dcg_graphs(), and genevalmag::Builder_graphs::save_dp_graphs().
.SS "void utilities::merge_graph (const Graph & graph1, const Graph & graph2, Graph & graph_merged)"Joins graph1 and graph2 in graph_merged. 
.PP
\fBParameters:\fP
.RS 4
\fIgraph1\fP 
.br
\fIgraph2\fP 
.br
\fIgraph_merged\fP Joins graph1 and graph2 in graph_merged. 
.RE
.PP

.PP
Definition at line 152 of file Utilities.cpp.
.PP
References return_vertex().
.PP
Referenced by genevalmag::Builder_graphs::combined_inf_contexts(), genevalmag::Builder_graphs::complete_dp_graphs(), genevalmag::Builder_graphs::compute_dcg(), and genevalmag::Builder_graphs::compute_down_graph().
.SS "void utilities::print_graph (const Graph & graph, const string path, const string name_file, const string name_graph, const string names[], string shape_vertex)"Prints a graph in a file .dot for generate image .spng. 
.PP
\fBParameters:\fP
.RS 4
\fIgraph\fP 
.br
\fIpath\fP 
.br
\fIname_file\fP 
.br
\fIname_graph\fP 
.br
\fInames\fP 
.br
\fIshape_vertex\fP Prints a graph in a file .dot for generate image .spng. 
.RE
.PP

.PP
Definition at line 102 of file Utilities.cpp.
.PP
Referenced by genevalmag::Builder_graphs::save_adp_graphs(), genevalmag::Builder_plans::save_all_plans(), genevalmag::Builder_plans::save_all_plans_project(), genevalmag::Builder_graphs::save_cyclic_graphs(), genevalmag::Builder_graphs::save_dcg_graphs(), genevalmag::Builder_graphs::save_down_graphs(), and genevalmag::Builder_graphs::save_dp_graphs().
.SS "void utilities::project_graph (const \fBSymbol\fP * symb, Graph & graph)"Projects a graph with only vertex that belongs to symbol 'symb'. Modifies the parameter 'graph'. 
.PP
\fBParameters:\fP
.RS 4
\fIsymb\fP 
.br
\fIgraph\fP Projects a graph with only vertex that belongs to symbol 'symb'. Modifies the parameter 'graph'. 
.RE
.PP

.PP
Definition at line 192 of file Utilities.cpp.
.PP
References genevalmag::Symbol::equals(), genevalmag::Expr_instance::get_num(), and genevalmag::Expr_instance::get_symb().
.PP
Referenced by genevalmag::Builder_graphs::compute_dcg(), and genevalmag::Builder_graphs::compute_down_graph().
.SS "Vertex utilities::return_vertex (const Graph & graph, const \fBExpr_leaf\fP * node)"Given a graph and node, returns the vertex descriptor of node in the graph. If not search it, so returns USHRT_MAX. 
.PP
\fBParameters:\fP
.RS 4
\fIgraph\fP 
.br
\fInode\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP
Given a graph and node, returns the vertex descriptor of node in the graph. If not search it, so returns USHRT_MAX. 
.PP
Definition at line 138 of file Utilities.cpp.
.PP
References genevalmag::Expr_leaf::equals().
.PP
Referenced by genevalmag::Builder_graphs::compute_attr_vertex(), genevalmag::Builder_graphs::compute_dependency_graphs(), genevalmag::Builder_plans::compute_order(), genevalmag::cycle_detector::examine_edge(), and merge_graph().
.SS "string utilities::write_inf_context (const vector< unsigned short > & context_vec)"Writes inferior contex in a string and returns it. 
.PP
\fBParameters:\fP
.RS 4
\fIcontext_vec\fP 
.RE
.PP
\fBReturns:\fP
.RS 4
.RE
.PP

.PP
Definition at line 236 of file Utilities.cpp.
.PP
Referenced by genevalmag::Builder_graphs::save_adp_graphs(), genevalmag::Builder_plans::save_all_plans(), genevalmag::Builder_plans::save_all_plans_project(), and genevalmag::Builder_graphs::save_cyclic_graphs().
.SH "Author"
.PP 
Generated automatically by Doxygen for maggen from the source code.


trunk/src/Util/Utilities.cpp
/**
  *  \file		Utilities.h
  *  \brief		Implementation of utilities module, where are methods and function used by many class.
  *  \date		09/03/2010
  *  \author	Kilmurray, Gerardo Luis <gerakilmurray@gmail.com>
  *  \author	Picco, Gonzalo Martin <gonzalopicco@gmail.com>
  */

#include <fstream>

#include <boost/graph/graphviz.hpp>
#include <boost/graph/transitive_closure.hpp>

#include "../../include/Util/Utilities.h"

using namespace std;
using namespace boost;
using namespace genevalmag;

namespace utilities
{

/**
  * Create the folder passed as parameter.
  */
bool create_folder(const string path)
{
	/* Create folder. */
	string command_mkdir_folder("mkdir -p ");
	command_mkdir_folder.append(path);
	if (system (command_mkdir_folder.c_str()) != 0)
	{
		return false;
	}
	return true;
}

/**
  * Remove and create the output folder of files .dot and .png.
  */
bool clean_output_folder(const string path)
{
	/* Remove the folder. */
	string command_rm_folder("rm -f -r ");
	command_rm_folder.append(path);
	if (system (command_rm_folder.c_str()) != 0)
	{
		return false;
	}
	/* Create folder. */
	return create_folder(path);
}

/**
  * Copies the static file to the generation path.
  */
bool copy_static_code(const string path_d, const string path_s)
{
	/* Copies files. */
	string command_cp_("cp -f ");
	command_cp_.append(path_s);
	command_cp_.append("lib/* ");
	command_cp_.append(path_d);

	if (system (command_cp_.c_str()) != 0)
	{
		return false;
	}
	return true;
}

/**
  * Generates the names of vertex. (The vertex's name is an instance).
  */
void generate_names_instance(const Graph &graph, string datas[], size_t size_d)
{
	property_map<Graph, vertex_data_t>::const_type props(get(vertex_data_t(), graph));

	for(size_t i(0); i < size_d; i++)
	{
		datas[i] = props[i]->to_string();
	}
}

/**
  * Generates the names of vertex for down graph. (The vertex's name is an attribute).
  */
void generate_names_attr(const Graph &graph, string datas[], size_t size_d)
{
	property_map<Graph, vertex_data_t>::const_type props(get(vertex_data_t(), graph));

	for(size_t i(0); i < size_d; i++)
	{
		/* The vertexs in a Down graph are ONLY Expr_instance. */
		datas[i] = ((Expr_instance*)props[i])->get_attr()->get_name();
	}
}

/**
  * Prints a graph in a file .dot for generate image .spng.
  */
void print_graph(const Graph &graph, const string path, const string name_file, const string name_graph, const string names[], string shape_vertex)
{
	static unsigned short num_file(0); /* For name of file png. */

	/* Create file dot. */
	string n_f(path);
	stringstream ins;
	ins << ++num_file;
	n_f.append(ins.str());

	n_f.append(name_file);

	n_f.append(".dot");

	/* Obtain of file of graphviz. */
	ofstream salida(n_f.c_str());

	map<string,string> graph_attr, vertex_attr, edge_attr;

	/* Sets the main characteristic of the graph. */
	graph_attr["label"]			= name_graph;
	graph_attr["rankdir"]		= "TD";
	graph_attr["ratio"]			= "compress";
	vertex_attr["shape"]		= shape_vertex;
	vertex_attr["color"] 		= "black";
	vertex_attr["style"] 		= "rounded,filled";
	vertex_attr["fillcolor"]	= "gray";

	/* Saves the dot file using the graphviz library. */
	write_graphviz(salida, graph, make_label_writer(names),default_writer(),make_graph_attributes_writer(graph_attr, vertex_attr,edge_attr));
}

/**
  * Given a graph and node, returns the vertex descriptor of node in the graph.
  * If not search it, so returns USHRT_MAX.
  */
Vertex return_vertex(const Graph &graph, const Expr_leaf *node)
{
	property_map<Graph, vertex_data_t>::const_type props(get(vertex_data_t(), graph));
	for(size_t i(0); i < num_vertices(graph); i++)
	{
		if (node->equals(props[i]))
			return i;
	}
	return USHRT_MAX;
}

/**
  * Joins graph1 and graph2 in graph_merged.
  */
void merge_graph(const Graph &graph1, const Graph &graph2, Graph &graph_merged)
{
	/* Cleans the result graph. */
	graph_merged.clear();

	/* Copies the graph1 to result graph. */
	graph_merged = graph1;

	/* Join the graph2 to result graph. */
	property_map<Graph, vertex_data_t>::const_type props(get(vertex_data_t(), graph2));
	property_map<Graph, vertex_data_t>::type props_merged(get(vertex_data_t(), graph_merged));
	/* Circle for vertices. */
	for(size_t i(0); i < num_vertices(graph2); i++)
	{
		Vertex vertex(return_vertex(graph_merged,props[i]));
		if (vertex == USHRT_MAX)
		{
			/* The vertex is new in the graph. */
			vertex = add_vertex(graph_merged);
			put(props_merged,vertex,props[i]);
		}
	}
	/* Cicle for edges. */
	graph_traits<Graph>::edge_iterator ei, ei_end;
	for (tie(ei,ei_end) = edges(graph2); ei != ei_end; ++ei)
	{
		Vertex source_vertex(source(*ei, graph2));
		Vertex target_vertex(target(*ei, graph2));

		Vertex desc_source(return_vertex(graph_merged,props[source_vertex]));
		Vertex desc_target(return_vertex(graph_merged,props[target_vertex]));

		add_edge(desc_source,desc_target,graph_merged);
	}
}

/**
  * Projects a graph with only vertex that belongs to symbol "symb".
  * Modifies the parameter "graph".
  */
void project_graph(const Symbol *symb, Graph &graph)
{
	/* Applies transitivity to graph with only nodes of symb. */
	warshall_transitive_closure(graph);

	property_map<Graph, vertex_data_t>::type props(get(vertex_data_t(), graph));
	/* Reduces the graph for symbol "symb". */
	for (size_t i(num_vertices(graph)); i > 0; i--)
	{
		const Expr_instance *ins(dynamic_cast<const Expr_instance*>(props[i-1]));
		if (!(ins) || !(ins->get_symb()->equals(*symb)) || !(ins->get_num() == 0))
		/* The node is a literal-node or is a node with symbol diferent that symb. */
		{
			clear_vertex(i-1, graph);
			remove_vertex(i-1, graph);
		}
	}
}

bool belong_index(const unsigned short &index, const vector<unsigned short> &vec)
{
	for(size_t i(0); i < vec.size(); i++)
	{
		if(index == vec[i])
		{
			return true;
		}
	}
	return false;
}

/**
  * Remove tabs and replace for spaces.
  */
string cleaning_tabs(const string str)
{
	string aux("");
	for(int i=0; str[i] != '\0'; i++)
	{
		aux.push_back((str[i] == '\t')? ' ': str[i]);
	}
	return aux;
}

string write_inf_context(const vector<unsigned short> &context_vec)
{
	string context;
	if (context_vec.size() == 1)
	{
		context.append(" hasn't an inferior context");
	}
	else
	{
		context.append(" with inferior context: ");
		for(size_t i(1); i < context_vec.size(); i++)
		{
			context.append(" R");
			stringstream key_rule;
			key_rule << context_vec[i];
			context.append(key_rule.str());
			if(i < context_vec.size() - 1)
			{
				context.append(" ,");
			}
		}
	}
	return context;
}

} /* end utilities */
